src/builder/QuarkBuilder.sol

```diff
@@ -128,9 +139,28 @@ contract QuarkBuilder {
                         revert TooManyBridgeOperations();
                     }
 
-                    uint256 amountToBridge = srcAccountBalances[j].balance >= amountLeftToBridge
-                        ? amountLeftToBridge
-                        : srcAccountBalances[j].balance;
+                    uint256 amountToBridge;
+                    // Handle differently if the transfer intent token is the payment token
+                    if (payment.isToken && Strings.stringEqIgnoreCase(payment.currency, transferIntent.assetSymbol)) {
+                        // Apply payment token logics to figure out the amount to bridge
+                        if (
+                            srcAccountBalances[j].balance
+                                >= amountLeftToBridge + PaymentInfo.findMaxCost(payment, srcChainAccounts.chainId)
+                        ) {
+                            amountToBridge = amountLeftToBridge;
+                        } else {
+                            amountToBridge = srcAccountBalances[j].balance
+                                - PaymentInfo.findMaxCost(payment, srcChainAccounts.chainId);
```

Comment:
I think this logic only makes sense if there is a single account on each chain. I think that's a fair assumption to make now, but let's leave a `NOTE:` detailing this assumption.